File: 2024/day11.py
#!/usr/bin/env python3
import math
from functools import cache

# Note: the `@cache` wrapper obscures the argument names from inline hints.

@cache
def blink(times: int, stone: int) -> int:
    if times == 0:
        return 1

    if stone == 0:
        result = blink(times - 1, 1)
    else:
        digit_count = int(math.log10(stone)) + 1
        if digit_count % 2 == 0:
            divisor = 10 ** (digit_count // 2)
            result = blink(times - 1, stone // divisor) + blink(times - 1, stone % divisor)
        else:
            result = blink(times - 1, stone * 2024)

    return result

def main() -> None:
    stones = list(map(int, input().strip().split()))
    for times in [25, 75]:
        total = 0
        for stone in stones:
            total += blink(times, stone)
        print(total)

if __name__ == "__main__":
    main()

File: 2024/day11.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

@cache = {}

# @param times [Integer] The number of times to blink.
# @param stone [Integer] The initial stone.
# @return [Integer] The number of stones after all the blinks.
def blink(times, stone)
  return 1 if times == 0

  state = [times, stone]

  @cache[state] ||= begin
    times -= 1
    if stone == 0
      blink(times, 1)
    elsif (digit_count = Math.log10(stone).to_i + 1).even?
      lhs, rhs = stone.divmod(10**(digit_count / 2))
      blink(times, lhs) + blink(times, rhs)
    else
      blink(times, stone * 2024)
    end
  end
end

stones = $stdin.gets.split.map(&:to_i)

[25, 75].each do |times|
  count = stones.sum { |stone| blink(times, stone) }
  puts count
end

#puts "Cache size: #{@cache.count}"
